New page: important

estelle opened the pull request:
<!-- 👀 Thanks for opening a PR! Read comments like this one to get your PR merged faster. -->
#### Summary
<!-- ✍️ In a sentence or two, describe your changes -->

#### Motivation
<!-- ❓ Why are you making this change? Help us understand how your changes help readers. -->

#### Supporting details
<!-- 🔗 Link to supporting information, such as bug trackers, source control, release notes, and vendor docs. -->

#### Related issues
<!-- 🔨 If applicable, use "Fixes #XYZ" -->

#### Metadata
<!-- ✅ Check a box if applicable, like this: [x]

This PR…
-->
- [X] Adds a new document
- [ ] Rewrites (or significantly expands) a document
- [ ] Fixes a typo, bug, or other error

<!-- 👷‍♀️ After submitting, review the results of the "Checks" tab! -->

estelle committed:
New page: important

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,90 @@
+---
+title: important
```
Drive-by comment:

Shouldn't we add the bang to the title? `!important`

estelle replied:
i was going to ask "how do you do that". If I include the ! I get an error.

Need to know:
* which, if any,  {{}} should be included for side nav.
* is there a way to change the URL and title to !important?

hamishwillee replied:
@estelle You might try quoting the title (untested): `title: '!important'`

No idea about the url.

OnkarRuikar replied:
> change the URL to !important

`!` is a valid character in URLs and file names

estelle committed:
New page: important

estelle committed:
rewrote it

estelle committed:
rewrote it

estelle committed:
Merge branch 'main' into important

OnkarRuikar commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,133 @@
+
+## Impact on the cascade
+
+When it comes to important declarations, the [cascade origin and layer orders](/en-US/docs/Web/CSS/Cascade) are reversed.  By default, declarations in an author (web developer) style sheets override declarations in a user’s style sheet, which override declarations in the user-agent’s default style sheet. 
```
```suggestion
When it comes to important declarations, the [cascade origin and layer orders](/en-US/docs/Web/CSS/Cascade) are reversed. By default, declarations in a website author's style sheets override declarations in a user’s style sheet, which override declarations in the user-agent’s default style sheet.
```

estelle commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,133 @@
+---
+title: important
```
```suggestion
title: '!important'
```

estelle committed:
Apply suggestions from code review

Co-authored-by: Onkar Ruikar <87750369+OnkarRuikar@users.noreply.github.com>

estelle committed:
updated best practices

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+# browser-compat: css.flag.important
+---
+
+The important flag, comprised of a `!` delimiter and `important` keyword, when added to a declaration in a selector block, makes the declaration an important declaration. 
```
```suggestion
A `!` delimiter followed by the `important` keyword sets the **important** flag on a declaration. The `!important` flag alters the rules selecting declarations inside the [cascade](/en-US/docs/Web/CSS/Cascade). A declaration that is not _important_ is called _normal_.
```

estelle replied:
still need to say "makes the declaration an important declaration."

teoli2003 replied:
Better?

```suggestion
A `!` delimiter followed by the `important` keyword marks the declaration as important. The `!important` flag alters the rules selecting declarations inside the [cascade](/en-US/docs/Web/CSS/Cascade). A declaration that is not _important_ is called _normal_.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+The important flag, comprised of a `!` delimiter and `important` keyword, when added to a declaration in a selector block, makes the declaration an important declaration. 
+
+## Description
+
```
```suggestion
```

teoli2003 replied:
I don't think this section heading is necessary. We always start reference articles with a description.

```suggestion
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+## Description
+
+To create an important declaration, add the "important flag", comprised of the delimiter token `!` followed by the `important` keyword, after the value in the declaration. While white space is allowed between the delimeter and the keyword, the important flag is generally written as `!important`, without any white space.
```
```suggestion
To mark a declaration important, add the _important flag_ (`!important`) after the value in the declaration. While white space is allowed between the delimiter and the keyword, the flag is generally written as `!important` without any white space.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
````diff
@@ -0,0 +1,135 @@
+}
+```
+
+The important flag must come after the property value pair declaration value, preceded by at least one space. The `!important` or `! important` must be the last tokens in the declaration. In other words, there can be white space and comments between the flag and the declarations ending semi-colon, but nothing else.  
````
```suggestion
The `!important` flag comes after the value of the property value pair declaration, preceded by at least one space. The `!important` flag must be the last token in the declaration. In other words, there can be white space and comments between the flag and the declaration's ending semi-colon, but nothing else.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+## Impact on the cascade
+
+When it comes to important declarations, the [cascade origin and layer orders](/en-US/docs/Web/CSS/Cascade) are reversed. By default, declarations in a website author's style sheets override declarations in a user’s style sheet, which override declarations in the user-agent’s default style sheet.
```
```suggestion
When it comes to `!important` declarations, the [cascade origin and layer orders](/en-US/docs/Web/CSS/Cascade) are reversed. Without the important flag, declarations in the author's style sheets override declarations in a user's style sheet, which override declarations in the user-agent's default style sheet.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+When it comes to important declarations, the [cascade origin and layer orders](/en-US/docs/Web/CSS/Cascade) are reversed. By default, declarations in a website author's style sheets override declarations in a user’s style sheet, which override declarations in the user-agent’s default style sheet.
+
+When a declaration is important, the order of precedence is reversed. The user-agent important declarations override all user important declarations. All user important declarations override all important author declarations. And all important declarations take precedence over all animations.  
```
```suggestion
When a declaration is `!important`, the order of precedence is reversed. Declarations marked as `!important` in the user-agent style sheets override all `!important` declarations in the user style sheets. Similarly, all `!important` declarations in the user style sheets override all `!important` declarations in the author's style sheets. Finally, all `!important` declarations take precedence over all animations.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+> **Note:** All important declarations take precedence over all animations. `!important` is not valid on within [@keyframes animation](/en-US/docs/Web/CSS/@keyframes) declarations.
+
+Reversing the precedence order for important declarations ensures users with special needs, such as personalized color schemes or large fonts, can override author styles when needed with important declarations. It also ensures malicious extensions can't override important user-agent styles, which might break functionality or negatively impact security. 
```
```suggestion
Reversing the precedence order for `!important` declarations ensures users with special needs, such as personalized color schemes or large fonts, can override author styles when needed by marking some declarations in their user's style sheet as `!important`. It also guarantees malicious extensions can't override `!important` user-agent styles, which might break functionality or negatively impact security.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+Reversing the precedence order for important declarations ensures users with special needs, such as personalized color schemes or large fonts, can override author styles when needed with important declarations. It also ensures malicious extensions can't override important user-agent styles, which might break functionality or negatively impact security. 
+
+Does anything have precedence over important declarations? Yes. [Transitions](/en-US/docs/Web/CSS/CSS_Transitions). CSS transitions are a way to control the speed at which the property changes from one value to another.  A transitioning property, while transitioning from one value to another, will not match a specific important declaration. 
```
```suggestion
Does anything have precedence over important declarations? Yes, [transitions](/en-US/docs/Web/CSS/CSS_Transitions). CSS transitions are a way to control the speed at which the property changes from one value to another. While transitioning from one value to another, a property will not match a specific important declaration.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
````diff
@@ -0,0 +1,135 @@
+}
+```
+
+In this example, the color and background color will transition to the hovered state over two seconds. Even though the default states are normal and hover states are important declarations, the transition does happen.
````
```suggestion
In this example, the `color` and `background-color` properties will transition to the hovered state over two seconds. Even though default states are normal declarations and hover states are `!important` declarations, the transition does happen.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+### Cascade layers
+
+Within each of the three origins - author, user, and user-agent - normal declarations in unlayered styles override layered style declarations, with the last declared having precedence over the layers declared before it. The order of precedence is reversed for important declarations, with important declarations in the first declared layer taking precedence over important declarations in the last d ...
```
```suggestion
Within each of the three origins for style sheets – author, user, and user-agent – normal declarations in unlayered styles override layered style declarations, with the last declared having precedence over the layers declared before it. Important declarations reverse the order of precedence: `!important` declarations in the first layer take precedence over `!important!` declarations in the next layer, and so on. Also, all the `!important` declarations have precedence over `!important` declarations made outside of any layer.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+### Inline styles
+
+When it comes to the [`style` attribute](/en-US/docs/Web/HTML/Global_attributes/style), inline normal styles take precedence over all normal declarations, no matter the origin and inline important styles take precedence over all other important author styles, no matter the layer, but are overridden by important user styles, important user-agent styles and transitions.
```
```suggestion
Inline styles are styles defined using the [`style`](/en-US/docs/Web/HTML/Global_attributes/style) attributes. They can also be normal or important. Inline _normal_ styles take precedence over all _normal_ declarations, no matter the origin. Inline _important_ styles take precedence over all other _important_ author styles, no matter the layer, but important styles from user's or user-agent's style sheets and transitions override them.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+### !important and specificity
+
+While `!important` is not part of determining specificity, it is related.  Important declarations override all other declarations from the same [origin and cascade layer](/en-US/docs/Web/CSS/Cascade). 
```
```suggestion
While `!important` is not part of determining specificity, it is related. Important declarations override all other declarations from the same [origin and cascade layer](/en-US/docs/Web/CSS/Cascade).
```

teoli2003 commented on files/en-us/web/css/important/index.md:
````diff
@@ -0,0 +1,135 @@
+}
+```
+
+This is a case of over-specifying a selector. No matter how high the selector  [specificity](/en-US/docs/Web/CSS/Specificity) match of a normal declaration, an important declaration from the same origin and cascade layer will always have precending.  In this case, the paragraph will always be red.
````
```suggestion
This example displays a case of over-specifying a selector. No matter how high the selector [specificity](/en-US/docs/Web/CSS/Specificity) matches a normal declaration, an `!important` declaration from the same source and cascade layer will always have precedence. In this case, the paragraph will always be red.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+This is a case of over-specifying a selector. No matter how high the selector  [specificity](/en-US/docs/Web/CSS/Specificity) match of a normal declaration, an important declaration from the same origin and cascade layer will always have precending.  In this case, the paragraph will always be red.
+
+When two important declarations are applied to the same element, with the same origin and layer, the declaration with a greater specificity will be applied.
```
```suggestion
When two `!important` declarations from the same origin and layer apply to the same element, browsers select and use the declaration with the highest specificity.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+## Impact on shorthand properties
+
+Declaring a shorthand property to be `!important` sets all of sub-properties `!important`. To two following selector style blocks are equivalent:
```
```suggestion
Declaring a shorthand property to be `!important` defines all of the longhand properties as`!important`. To two following selector style blocks are equivalent:
```

teoli2003 commented on files/en-us/web/css/important/index.md:
````diff
@@ -0,0 +1,135 @@
+}
+ ``` 
+
+This is one of the several reasons avoiding `!important` is generally recommended.
````
```suggestion
This example shows one of the several reasons avoiding `!important` is generally recommended.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+## Best practices
+
+Avoid using `!important` to override specificity. When intentionally creating important declarations for UI requirements, comment your CSS to explain to maintainers why a feature must not be overridden.
```
```suggestion
Avoid using `!important` to override specificity. When intentionally creating `!important` declarations for UI requirements, comment in your CSS code to explain to maintainers why they should not override that feature.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+Avoid using `!important` to override specificity. When intentionally creating important declarations for UI requirements, comment your CSS to explain to maintainers why a feature must not be overridden.
+
+If working to override styles with high scecificity that are not under your control, such as styles in a 3rd party plugin declared with an [id selector](/en-US/docs/Web/CSS/ID_selectors), consider importing the 3rd party stylesheet script into a [named or anonymous layer](/en-US/docs/Web/CSS/@layer) as your first cascade layer, instead of using `!important`. As long as the external styles do not  ...
```
```suggestion
Even when working to override high-specificity styles not under your control, such as styles in a 3rd party plugin declared with an [id selector](/en-US/docs/Web/CSS/ID_selectors), you don't need to use `!important`. Consider instead importing the 3rd party stylesheet script into a [named or anonymous layer](/en-US/docs/Web/CSS/@layer) as your first cascade layer, instead of using `!important`. As long as the external styles do not include `!important` declarations, your styles will take precedence over the widget styles, no matter the specificity.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+If working to override styles with high scecificity that are not under your control, such as styles in a 3rd party plugin declared with an [id selector](/en-US/docs/Web/CSS/ID_selectors), consider importing the 3rd party stylesheet script into a [named or anonymous layer](/en-US/docs/Web/CSS/@layer) as your first cascade layer, instead of using `!important`. As long as the external styles do not  ...
+
+If needing to override external stylesheet containing important declarations, create a cascade layer of importnant overrides and declare that layer first. 
```
```suggestion
If you need to override an external stylesheet containing `!important` declarations, create a cascade layer containing the needed overrides, and declare that layer first.
```

teoli2003 commented on files/en-us/web/css/important/index.md:
```diff
@@ -0,0 +1,135 @@
+
+### Accessibility
+
+Important styles from a user stylesheet take precedence over author important declarations, meaning adding an important flag to a site's styles will not prevent individual users with special requirements, such as large fonts, from being able to override your styles with their own important styles.
```
```suggestion
Important styles from a user stylesheet take precedence over the author style sheet's `!important` declarations, meaning adding an `!important` flag to a site's styles will not prevent individual users with special requirements, such as large fonts, from being able to override your styles by adding`!important` styles in their own user's style sheet.
```

teoli2003 reviewed (changes requested):
I learned a lot reading this. Amazing work. I found some sentences difficult to understand, at least for a non-native speaker, so I have tried to shorten them and to use a simpler grammar when possible.

I probably added some new grammar errors, so you'll likely need an extra pass.

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
normalization of keyterm usage.

changed some `!important` to "important", and few "important" to `!important`, and two grammar fixes

OnkarRuikar commented on files/en-us/web/css/important/index.md:
````diff
@@ -0,0 +1,133 @@
+}
+```
+
+This is a case of over-specifying a selector. No matter how high the selector  [specificity](/en-US/docs/Web/CSS/Specificity) match of a normal declaration, an important declaration from the same origin and cascade layer will always have precending.  In this case, the paragraph will always be red.
````
typo
```suggestion
This is a case of over-specifying a selector. No matter how high the selector [specificity](/en-US/docs/Web/CSS/Specificity) match of a normal declaration, an important declaration from the same origin and cascade layer will always have precedence. In this case, the paragraph will always be red.
```

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Update files/en-us/web/css/important/index.md

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
Apply suggestions from code review

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

estelle committed:
redo normalization

teoli2003 reviewed (approved):
👍 Great article.
